astronomy/date_body.hpp

```diff
@@ -1041,6 +1067,148 @@ constexpr bool IsValidStretchyUTC(DateTime const& utc) {
           IsValidPositiveStretchyUTCLeap(utc, "1972-01-01"_Date, 107.7580));
 }
 
+// Utilities for UT1.
+
+// An entry in the Experimental EOP C02 time series; represents UT1 - TAI at the
+// given |ut1_mjd|.
+struct ExperimentalEOPC02Entry {
+  constexpr ExperimentalEOPC02Entry(double const ut1_mjd,
+                                    quantities::Time const ut1_minus_tai);
+
+  double const ut1_mjd;
+  quantities::Time const ut1_minus_tai;
+};
+
+constexpr ExperimentalEOPC02Entry::ExperimentalEOPC02Entry(
+    double const ut1_mjd,
+    quantities::Time const ut1_minus_tai)
+    : ut1_mjd(ut1_mjd),
+      ut1_minus_tai(ut1_minus_tai) {}
+
+// An entry in the EOP (IERS) 08 C04 time series; represents UT1 - UTC at
+// 00:00:00 on the given |utc_date|.  The date is given as an integer of the
+// form YYYY'MM'DD, which is then interpreted on demand, in order to limit the
+// number of constexpr steps.
+struct EOPC04Entry {
+  constexpr EOPC04Entry(int const utc_date,
+                        quantities::Time const& ut1_minus_utc);
+
+  constexpr DateTime utc() const;
+  constexpr quantities::Time ut1() const;
+  constexpr quantities::Time ut1_minus_tai() const;
+
+  int const utc_date;
+  quantities::Time const ut1_minus_utc;
+};
+
+constexpr EOPC04Entry::EOPC04Entry(
+    int const utc_date,
+    quantities::Time const& ut1_minus_utc)
+    : utc_date(utc_date),
+      ut1_minus_utc(ut1_minus_utc) {}
+
+constexpr DateTime EOPC04Entry::utc() const {
+  return DateTime::BeginningOfDay(Date::YYYYMMDD(utc_date));
+}
+
+constexpr quantities::Time EOPC04Entry::ut1() const {
+  return TimeScale(utc()) + ut1_minus_utc;
+}
+
+constexpr quantities::Time EOPC04Entry::ut1_minus_tai() const {
+  return utc().date().year() >= 1972
+             ? ut1_minus_utc + ModernUTCMinusTAI(utc().date())
+             : ut1_minus_utc - TAIMinusStretchyUTC(utc());
+}
+
+// NOTE(egg): these have to be defined here, because they depend on internal
```

Comment:
Moving the EOP entries to their own file would be very difficult, since their methods depend on internals of the UTC interpretation (`TAIMinusStretchyUTC` and the like). On the other hand, it is reasonably easy to split Date, Time, DateTime, and the parsing of their literals from the interpretation according to various time scales.